Exclude zero residuals in forecast uncertainty sampling

moritzpschwarz opened the pull request:
Updated the forecast_model function to exclude zero residuals when sampling for uncertainty, preventing underestimation due to IIS-induced zeros. Updated test snapshot images to reflect changes in forecast outputs.

Closes #17

moritzpschwarz committed:
Exclude zero residuals in forecast uncertainty sampling

Updated the forecast_model function to exclude zero residuals when sampling for uncertainty, preventing underestimation due to IIS-induced zeros. Updated test snapshot images to reflect changes in forecast outputs.

jkurle committed:
change how zero res are identified and styler

- make zero decision base don machine floating imprecision
- same as testthat_tolerance() or all.equal()

jkurle commented:
Four points:

1) I suggested an alternative for identifying zero residuals, which is based on what base R and testthat seem to do: use the user's machine precision for floating numbers and use that as a tolerance to tell values from zero (or one another) apart.
2) Alternative idea: what if the fit of the model is simply very good and that's why we get small residuals? The size of residuals is also a matter of the scale of the dependent variable...should we instead extract the IIS-names of the isat object and exclude those residuals? Sounds like the most robust approach to me.
3) I was unable to update the tests in case it made any difference. The tests don't run reliably on my machine, we need to fix this! (renv?)
4) I was unable to make my commit without the styler changes. I'm not sure why the function wasn't already styled since we have the pre-commit hook. However, the version that was on the branch raised loads of issues, so I had to update it. I hope that this does not raise further issues on your end. We should also discuss this as a workflow. If not all files are currently properly styled (including mine, I definitely did not use the styler at the beginning, so if I have not made any commits to these files since the styler was activated, they would still be wrongly styled), we should do a round of commits where we do only the styler changes, so we don't mix them with more substantive changes - otherwise we clutter the actual changes.

moritzpschwarz commented:
> Four points:
> 
> 1. I suggested an alternative for identifying zero residuals, which is based on what base R and testthat seem to do: use the user's machine precision for floating numbers and use that as a tolerance to tell values from zero (or one another) apart.
> 2. Alternative idea: what if the fit of the model is simply very good and that's why we get small residuals? The size of residuals is also a matter of the scale of the dependent variable...should we instead extract the IIS-names of the isat object and exclude those residuals? Sounds like the most robust approach to me.

Ok, this sounds good to me - I will implement that it only takes the IIS and removes them. 

> 3. I was unable to update the tests in case it made any difference. The tests don't run reliably on my machine, we need to fix this! (renv?)

Will look at specifically.

> 4. I was unable to make my commit without the styler changes. I'm not sure why the function wasn't already styled since we have the pre-commit hook. However, the version that was on the branch raised loads of issues, so I had to update it. I hope that this does not raise further issues on your end. We should also discuss this as a workflow. If not all files are currently properly styled (including mine, I definitely did not use the styler at the beginning, so if I have not made any commits to these files since the styler was activated, they would still be wrongly styled), we should do a round of commits where we do only the styler changes, so we don't mix them with more substantive changes - otherwise we clutter the actual changes.

Will discuss separately

moritzpschwarz committed:
Exclude IIS residuals from uncertainty sampling

Updated the residual sampling logic in forecast_model to exclude residuals corresponding to IIS indices using setdiff, rather than filtering by machine precision. This aims to more accurately represent uncertainty by omitting residuals artificially set to zero by IIS.

moritzpschwarz committed:
Handle edge case with all IIS in residuals sampling

Adds a check to ensure that if all residuals are saturated (i.e., all are IIS), the uncertainty draws default to zero. This prevents errors in rare edge cases where no non-IIS residuals are available for sampling.

moritzpschwarz commented on R/forecast_model.R:
```diff
@@ -161,7 +161,9 @@ forecast_model <- function(model,
 
       # make samples from the model residuals and add them to the mean prediction
-      res_draws <- sample(as.numeric(isat_obj$residuals), size = uncertainty_sample * n.ahead, replace = TRUE)
+      res_nozero <- as.numeric(isat_obj$residuals)[round(as.numeric(isat_obj$residuals),10) != 0] # exclude 0 residuals (due to IIS) to not underestimate uncertainty
```
Added an edge case handling

moritzpschwarz commented:
Alright @jkurle, I think it's ready now to merge.

jkurle commented on R/forecast_model.R:
```diff
@@ -166,8 +166,10 @@ forecast_model <- function(model,
+      #tolerance <- sqrt(.Machine$double.eps)
+      #res_nozero <- as.numeric(isat_obj$residuals)[abs(as.numeric(isat_obj$residuals)) > tolerance] # exclude 0 residuals (due to IIS) to not underestimate uncertainty
+
+      res_nozero <- setdiff(isat_obj$residuals, isat_obj$residuals[gets::isatdates(isat_obj)$iis$index])
```
Does

`res_nozero <- isat_obj$residuals[-gets::isatdates(isat_obj)$iis$index]`

work? Your code probably works but I am not sure whether it is "stable"/"safe" to do a set comparison between numbers with many digits. So rather than taking the full set and subtracting the residual values that are from IIS, I would just directly filer out the IIS residuals by index.

moritzpschwarz replied:
Sure, happy to adjust to that - I think even with the setdiff it would have resulted in the same result, but now changed to your suggestion!

moritzpschwarz committed:
Change residuals subsetting in forecast_model

Replaces setdiff with negative indexing to exclude IIS indices from residuals.

moritzpschwarz committed:
Handle empty or missing IIS indices in residual exclusion

Adds a check to ensure IIS indices exist and are numeric before excluding them from residuals. This prevents errors when IIS indices are missing or empty, improving robustness in edge cases.

moritzpschwarz committed:
Fix type issue in residual sampling for forecast

Wraps res_nozero with as.numeric in the sample() call to ensure correct type is used when drawing residuals for uncertainty estimation.

jkurle commented:
great, looks good to go!
